File: lib/rgfa/graph_operations/linear_paths.rb
require_relative "../segment_end"
require_relative "./redundant_linear_paths"

#
# Methods for the RGFA class, which allow to find and merge linear paths.
#
# @tested_in api_linear_paths
#
module RGFA::GraphOperations::LinearPaths

  require "set"

  #
  # Find a path without branches.
  #
  # The path must include +s+ and exclude all segments in the +exclude+ set. Any
  # segment used in the returned path will be added to the +exclude+ set.
  #
  # @param s [String,Symbol,RGFA::Line::Segment::GFA1,RGFA::Line::Segment::GFA2]
  #   a segment name or instance
  # @param exclude [Set<String>] a set of segment names to exclude from the path
  # @return [Array<RGFA::SegmentEnd>]
  #
  def linear_path(s, exclude = Set.new)
    cs = segment(s).connectivity
    s = s.to_sym
    segpath = RGFA::SegmentEndsPath.new()
    [:L, :R].each_with_index do |et, i|
      if cs[i] == 1
        exclude << s
        segpath.pop
        segpath += traverse_linear_path(RGFA::SegmentEnd.new(s, et), exclude)
      end
    end
    return segpath
  end

  # Find all unbranched paths in the graph.
  #
  # @param redundant [Boolean]
  #     if true, junctions are added to paths ends, also if the same end of the
  #     junction node has also additional links
  #
  # @return [Array<Array<RGFA::SegmentEnd,Boolean>>] the booleans are added
  #           if the +redundant+ parameter is set, as first and last element
  #           (they are set if the first / last segment are junctions shall
  #           be duplicated when merging paths)
  def linear_paths(redundant = false)
    exclude = Set.new
    junction_exclude = Set.new if redundant
    retval = []
    segnames = segment_names
    progress_log_init(:linear_paths, "segments", segnames.size,
      "Detect linear paths (#{segnames.size} segments)")  if @progress
    segnames.each do |sn|
      progress_log(:linear_paths) if @progress
      next if exclude.include?(sn)
      lp = linear_path(sn, exclude)
      if !redundant
        retval << lp if lp.size > 1
      else
        if lp.empty?
          # add paths from junction to junction neighbours
          retval += junction_junction_paths(sn, junction_exclude)
        else
          # extend linear path to junction neighbours
          extend_linear_path_to_junctions(lp)
          retval << lp
        end
      end
    end
    progress_log_end(:linear_paths)
    return retval
  end

  # Merge a linear path, i.e. a path of segments without extra-branches
  # @!macro [new] merge_lim
  #   Limitations: all containments und paths involving merged segments are
  #   deleted.
  #
  # @param segpath [Array<RGFA::SegmentEnd,Boolean>] a linear path, such as that
  #   retrieved by {#linear_path}
  # @!macro [new] merge_options
  #   @param options [Hash] optional keyword arguments
  #   @option options [String, :short, nil] :merged_name (nil)
  #     if nil, the merged_name is automatically computed; if :short,
  #     a name is computed starting with "merged1" and calling next until
  #     an available name is founf; if String, the name to use
  #   @option options [Boolean] :cut_counts (false)
  #     if true, total count in merged segment m, composed of segments
  #     s of set S is multiplied by the factor Sum(|s in S|)/|m|
  #   @option options [Boolean] :redundant (false)
  #     if true, junctions are output multiple times, merging them with
  #     all paths which start or end in them; default: they are
  #     merged only with the path on the side which has one link only,
  #     if any, and left as separate nodes otherwise
  #   @option options [symbol] :jntag (:jn)
  #     specify temporary optional tag used for storing information about
  #     junctions used by the merged_linear_paths method when deleting
  #     them; it should be a tag which is otherwise not present in the segments
  #
  # @return [RGFA] self
  # @see #merge_linear_paths
  def merge_linear_path(segpath, **options)
    return if segpath.size < 2
    if options[:redundant] and ((segpath[0] == true) or (segpath[0] == false))
      first_redundant = segpath.shift
      last_redundant = segpath.pop
    else
      first_redundant, last_redundant = false
    end
    segpath.map!{|se|se.to_segment_end}
    merged, first_reversed, last_reversed =
      create_merged_segment(segpath, options)
    self << merged
    if first_redundant
      link_duplicated_first(merged,
                            segment(segpath.first.to_segment_end.segment),
                            first_reversed, options[:jntag])
    else
      link_merged(merged.name, segpath.first.to_segment_end.invert,
                  first_reversed)
    end
    if last_redundant
      link_duplicated_last(merged, segment(segpath.last.segment),
                           last_reversed, options[:jntag])
    else
      link_merged(merged.name, segpath.last, last_reversed)
    end
    idx1 = first_redundant ? 1 : 0
    idx2 = last_redundant ? -2 : -1
    segpath[idx1..idx2].each do |sn_et|
      segment(sn_et.segment).disconnect
      progress_log(:merge_linear_paths, 0.05) if @progress
    end
    self
  end

  # Merge all linear paths in the graph, i.e.
  # paths of segments without extra-branches
  # @!macro merge_lim
  # @!macro merge_options
  #
  # @return [RGFA] self
  def merge_linear_paths(**options)
    paths = linear_paths(options[:redundant])
    psize = paths.flatten.size / 2
    progress_log_init(:merge_linear_paths, "segments", psize,
      "Merge #{paths.size} linear paths (#{psize} segments)") if @progress
    paths.each do |path|
      merge_linear_path(path, **options)
    end
    progress_log_end(:merge_linear_paths)
    remove_junctions(options[:jntag]) if options[:redundant]
    self
  end

  private

  # Traverse the links, starting from the segment end +segment_end+.
  #
  # If any segment is found during traversal whose name is included in +exclude+
  # the traversing is interrupted. The +exclude+ set is updated, so that
  # circular paths are avoided. The starting segment is not added to the set.
  #
  def traverse_linear_path(segment_end, exclude)
    list = RGFA::SegmentEndsPath.new()
    current = segment_end.to_segment_end
    current.segment = segment(current.segment)
    loop do
      after  = current.segment.dovetails(current.end_type)
      before = current.segment.dovetails(current.end_type.invert)
      if (before.size == 1 and after.size == 1) or list.empty?
        list << [current.name, current.end_type]
        exclude << current.name
        l = after.first
        current = l.other_end(current).invert
        break if exclude.include?(current.name)
      elsif before.size == 1
        list << [current.name, current.end_type]
        exclude << current.name
        break
      else
        break
      end
    end
    return segment_end.end_type == :L ? list.reverse : list
  end

  def sum_of_counts(segpath, multfactor = 1)
    retval = {}
    segs = segpath.map {|sn_et|segment!(sn_et.segment)}
    [:KC, :RC, :FC].each do |count_tag|
      segs.each do |s|
        if s.tagnames.include?(count_tag)
          retval[count_tag] ||= 0
          retval[count_tag] += s.get(count_tag)
        end
      end
      if retval[count_tag]
        retval[count_tag] = (retval[count_tag] * multfactor).to_i
      end
    end
    return retval
  end

  def reverse_segment_name(name, separator)
    name.to_s.split(separator).map do |part|
      openp = part[0] == "("
      part = part[1..-1] if openp
      closep = part[-1] == ")"
      part = part[0..-2] if closep
      part = (part[-1] == "^") ? part[0..-2] : part+"^"
      part += ")" if openp
      part = "(#{part}" if closep
      part
    end.reverse.join(separator)
  end

  def reverse_pos_array(pos_array, lastpos)
    return nil if pos_array.nil? or lastpos.nil?
    pos_array.map {|pos| lastpos - pos + 1}.reverse
  end

  def add_segment_to_merged(merged, segment, reversed, cut, init, options)
    s = (reversed ? segment.sequence.rc[cut..-1] : segment.sequence[cut..-1])
    if init
      merged.sequence = s
      merged.name = (options[:merged_name].nil? ?
                     segment.name : options[:merged_name])
      merged.LN = segment.LN
    else
      (segment.sequence.placeholder?) ?
        (merged.sequence = RGFA::Placeholder.new) :
        (merged.sequence += s)
      if options[:merged_name].nil?
        merged.name = "#{merged.name}_#{segment.name}"
      end
      if merged.LN
        segment.LN ? merged.LN += (segment.LN - cut)
                   : merged.LN = nil
      end
    end
  end

  def create_merged_segment(segpath, options)
    merged = segment!(segpath.first.segment).clone
    merged.set(options[:jntag] || :jn, nil)
    total_cut = 0
    a = segpath.first
    first_reversed = (a.end_type == :L)
    last_reversed = nil
    if options[:merged_name] == :short
      forbidden = (segment_names + path_names)
      options[:merged_name] = "merged1"
      while forbidden.include?(options[:merged_name])
        options[:merged_name] = options[:merged_name].next
      end
    end
    add_segment_to_merged(merged, segment(a.segment), first_reversed, 0, true,
                          options)
    progress_log(:merge_linear_paths, 0.95) if @progress
    (segpath.size-1).times do |i|
      b = segpath[i+1].to_segment_end.invert
      ls = segment(a.segment).end_relations(a.end_type, b, :dovetails)
      if ls.size != 1
        msg = "A single link was expected between #{a} and #{b}, "+
          "#{ls.size} were found"
        msg << ":\n#{l.map(&:to_s).join("\n")}" if ls.size > 0
        raise RGFA::ValueError, msg
      end
      l = ls[0]
      if l.overlap == []
        cut = 0
      elsif l.overlap.all?{|op|[:M, :"="].include?(op.code)}
        cut = l.overlap.map(&:len).inject(:+)
      else
        raise RGFA::ValueError,
          "Merging is only allowed if all operations are M/="
      end
      total_cut += cut
      last_reversed = (b.end_type == :R)
      add_segment_to_merged(merged, segment(b.segment), last_reversed, cut,
                            false, options)
      a = b.to_segment_end.invert
      if @progress
        progress_log(:merge_linear_paths, 0.95)
      end
    end
    if !merged.sequence.placeholder?
      if merged.LN.nil?
        merged.LN = merged.sequence.length
      elsif @vlevel >= 1 and merged.LN != merged.sequence.length
        # XXX
        raise RGFA::InconsistencyError,
              "Computed sequence length #{merged.sequence.length} "+
              "and computed LN #{merged.LN} differ"
      end
    end
    if merged.LN.nil?
      [:KC, :RC, :FC].each {|count_tag| merged.set(count_tag, nil)}
    else
      sum_of_counts(segpath, (options[:cut_counts] ?
                              merged.LN.to_f / (total_cut+merged.LN) : 1)).
          each do |count_tag, count|
        merged.set(count_tag, count)
      end
    end
    return merged, first_reversed, last_reversed
  end

  def link_merged(merged_name, segment_end, reversed)
    segment(segment_end.segment).dovetails(
        segment_end.end_type).each do |l|
      l2 = l.clone
      if l2.to == segment_end.segment
        l2.to = merged_name
        if reversed
          l2.to_orient = l2.to_orient.invert
        end
      else
        l2.from = merged_name
        if reversed
          l2.from_orient = l2.from_orient.invert
        end
      end
      self << l2
    end
  end

  include RGFA::GraphOperations::RedundantLinearPaths

end

File: gfapy/graph_operations/linear_paths.py
import gfapy

class LinearPaths:

  def linear_path(self, segment, exclude = None):
    """Find a linear path which contains the specified segment

    Parameters:
      segment (str, Line): the segment to analyse
      exclude : (API private)
    """
    if isinstance(segment, gfapy.Line):
      segment_name = segment.name
    else:
      segment_name = segment
      segment = self.segment(segment_name)
    cs = segment._connectivity()
    if exclude is None:
      exclude = set()
    segpath = gfapy.SegmentEndsPath()
    for i, et in enumerate(["L", "R"]):
      if cs[i] == 1:
        exclude.add(segment_name)
        if len(segpath) > 0:
          segpath.pop()
        segpath += self.__traverse_linear_path(
            gfapy.SegmentEnd(segment, et), exclude)
    return segpath

  def linear_paths(self, redundant_junctions=False):
    """Find linear paths of dovetail overlaps connecting segments.

    Parameters:
      redundant_junctions (bool): output the junction segments at the
        end of each path which involves them; this mimics the construction
        of contigs in string graph assemblers Readjoiner and SGA; default: False
    """
    exclude = set()
    if redundant_junctions:
      junction_exclude = set()
    retval = []
    segnames = self.segment_names
    if self._progress:
      self._progress_log_init("linear_paths", "segments", len(segnames),
          "Detect linear paths ({})".format(len(segnames)))
    for sn in segnames:
      if self._progress:
        self._progress_log("linear_paths")
      if sn in exclude:
        continue
      lp = self.linear_path(sn, exclude)
      if not redundant_junctions:
        if len(lp) > 1:
          retval.append(lp)
      else:
        if lp:
          self._extend_linear_path_to_junctions(lp)
          retval.append(lp)
        else:
          retval += self._junction_junction_paths(sn, junction_exclude)
    if self._progress:
      self._progress_log_end("linear_paths")
    return retval

  def merge_linear_path(self, segpath, redundant_junctions=False, jntag="jn",
                        enable_tracking=False, merged_name=None,
                        cut_counts=False):
    """Merge a specified linear path of dovetail overlaps connecting segments.

    Note:
      for the parameter usage, see merge_linear_paths();
      the only difference is that merged_name can be set to a string (different
      from 'short'), which will be used as a name for the merged segment.
    """
    if len(segpath) < 2:
      return self
    if segpath[0] in [True, False]:
      first_redundant = segpath.pop(0)
      last_redundant = segpath.pop()
    else:
      first_redundant = False
      last_redundant = False
    segpath = [gfapy.SegmentEnd(s) for s in segpath]
    merged, first_reversed, last_reversed = \
        self.__create_merged_segment(segpath,
            redundant_junctions=redundant_junctions, jntag=jntag,
            merged_name=merged_name,cut_counts=cut_counts,
            enable_tracking=enable_tracking)
    self.append(merged)
    if first_redundant:
      self._link_duplicated_first(merged, self.segment(segpath[0].segment),
                                   first_reversed, jntag)
    else:
      self.__link_merged(merged.name, segpath[0].inverted(), first_reversed)
    if last_redundant:
      self._link_duplicated_last(merged, self.segment(segpath[-1].segment),
                                  last_reversed, jntag)
    else:
      self.__link_merged(merged.name, segpath[-1], last_reversed)
    idx1 = 1 if first_redundant else 0
    idx2 = -1 if last_redundant else None
    for sn_et in segpath[idx1:idx2]:
      self.segment(sn_et.segment).disconnect()
      if self._progress:
        self._progress_log("merge_linear_paths", 0.05)
    return self

  def merge_linear_paths(self, redundant_junctions=False, jntag="jn",
                        merged_name=None, enable_tracking=False,
                        cut_counts=False):
    """Find and merge linear paths of dovetail overlaps connecting segments.

    Note:
      Besides obviously the dovetail overlaps, all lines refererring to the
      merged segments (containments, internal edges, paths, sets, fragments,
      gaps) are removed from the Gfa instance.

    Parameters:
      merged_name (str): if 'short', then a name is computed using an unused
        integer; otherwise the name is computed using a combination of the
        names of the merged segments, separated by an underscore
      cut_counts (bool): if True, the total count in merged segment m,
         composed of segments s of set S is multiplied by the factor
         ``Sum(|s in S|)/|m|``
      enable_tracking: if True, tracking information is added as follows;
        the name of the component segments is stored in the ``or`` tag (or the
        content of their ``or`` tag, instead of the name, if any) and their
        starting positions is stored in the ``mp`` tag; the ``rn`` tag, used
        for storing possibe inversion positions by the random orientation
        methods of this library, is inherited and the positions updated;
        unless merged_name is set to 'short', the computation of the merged
        name is enhanced, in that reverse complement components are suffixed
        with ``^`` and parenthesis added by the random orientation methods of
        this library are inherited
      redundant_junctions (bool): output the junction segments at the
        end of each path which involves them; this mimics the construction
        of contigs in string graph assemblers Readjoiner and SGA; default: False
      jntag (str) : the tag to use for the temporary storage of junction
        information, if the redundant_junctions flag is set (default: jn)
    """
    paths = self.linear_paths(redundant_junctions)
    if self._progress:
      psize = sum([len(path) for path in paths])
      self._progress_log_init("merge_linear_paths", "segments", psize,
          "Merge {} linear paths ".format(len(paths))+
          "({} segments)".format(psize))
    for path in paths:
      self.merge_linear_path(path, redundant_junctions=redundant_junctions,
                             jntag=jntag, merged_name=merged_name,
                             cut_counts=cut_counts,
                             enable_tracking=enable_tracking)
    if self._progress:
      self._progress_log_end("merge_linear_paths")
    if redundant_junctions:
      self._remove_junctions(jntag)
    return self

  def __traverse_linear_path(self, segment_end, exclude):
    lst = gfapy.SegmentEndsPath()
    current = gfapy.SegmentEnd(segment_end)
    current.segment = self.segment(current.segment)
    while True:
      after = current.segment.dovetails_of_end(current.end_type)
      before = current.segment.dovetails_of_end(gfapy.invert(current.end_type))
      if (len(before) == 1 and len(after) == 1) or not lst:
        lst.append(gfapy.SegmentEnd(current.name, current.end_type))
        exclude.add(current.name)
        current = after[0].other_end(current).inverted()
        if current.name in exclude:
          break
      elif len(before) == 1:
        lst.append(gfapy.SegmentEnd(current.name, current.end_type))
        exclude.add(current.name)
        break
      else:
        break
    if segment_end.end_type == "L":
      return list(reversed(lst))
    else:
      return lst

  def __sum_of_counts(self, segpath, multfactor = 1):
    retval = {}
    segs = [self.try_get_segment(sn_et.segment) for sn_et in segpath]
    for count_tag in ["KC","RC","FC"]:
      for s in segs:
        if count_tag in s.tagnames:
          if count_tag not in retval:
            retval[count_tag] = 0
          retval[count_tag] += int(retval[count_tag]*multfactor)
    return retval

  def _add_segment_to_merged(self, merged, segment, is_reversed, cut, init,
                             enable_tracking=False, merged_name=None):
    n = segment.name
    if is_reversed:
      s = gfapy.sequence.rc(segment.sequence)[cut:]
      if enable_tracking:
        n = self._reverse_segment_name(segment.name, "_")
        rn = self._reverse_pos_array(segment.rn, segment.LN)
        mp = self._reverse_pos_array(segment.mp, segment.LN)
    else:
      s = segment.sequence[cut:]
      if enable_tracking:
        rn = segment.rn
        mp = segment.mp
    if enable_tracking:
      if not mp and segment.LN:
        mp = [1, segment.LN]
      if segment.get("or") is None:
        o = n
      elif is_reversed:
        o = self._reverse_segment_name(segment.get("or"), ",")
      else:
        o = segment.get("or")
    if init:
      merged.sequence = [s]
      if merged_name:
        merged.name = [merged_name]
      else:
        merged.name = [n]
      merged.LN = segment.LN
      if enable_tracking:
        merged.rn = rn
        merged.set("or",[o])
        merged.mp = mp
    else:
      if gfapy.is_placeholder(segment.sequence):
        merged.sequence = gfapy.Placeholder()
      else:
        merged.sequence.append(s)
      if not merged_name:
        merged.name.append(n)
      if merged.LN:
        if enable_tracking:
          if rn:
            rn = [pos - cut + merged.LN for pos in rn]
            if not merged.rn:
              merged.rn = rn
            else:
              merged.rn += rn
          if mp and merged.mp:
            merged.mp += [pos - cut + merged.LN for pos in mp]
        if segment.LN:
          merged.LN += (segment.LN - cut)
        else:
          merged.LN = None
      elif enable_tracking:
        merged.mp = None
      if enable_tracking:
        if not merged.get("or"):
          merged.set("or", [o])
        else:
          merged.get("or").append(o)

  @staticmethod
  def _reverse_segment_name(name, separator):
    retval = []
    for part in name.split(separator):
      has_openp = part[0] == "("
      has_closep = part[-1] == ")"
      if has_openp:
        part = part[1:-2]
      if has_closep:
        part = part[:-1]
      if part[-1] == "^":
        part = part[:-1]
      else:
        part+="^"
      if has_openp:
        part+=")"
      if has_closep:
        part+="("+part
      retval.append(part)
    return separator.join(reversed(retval))

  @staticmethod
  def _reverse_pos_array(pos_array, lastpos):
    if pos_array is None or lastpos is None:
      return None
    else:
      return [lastpos-pos+1 for pos in pos_array].reverse()

  def __create_merged_segment(self, segpath, redundant_junctions=False,
      jntag="jn", merged_name=None, enable_tracking=False, cut_counts=False):
    merged = self.try_get_segment(segpath[0].segment).clone()
    merged.set(jntag, None)
    merged_vlevel = merged.vlevel
    merged.vlevel = 0
    total_cut = 0
    a = segpath[0]
    first_reversed = (a.end_type == "L")
    last_reversed = None
    if merged_name == "short":
      merged_name = self.unused_name()
    self._add_segment_to_merged(merged, self.segment(a.segment),
        first_reversed, 0, True, enable_tracking=enable_tracking,
        merged_name=merged_name)
    if self._progress:
      self._progress_log("merge_linear_paths", 0.95)
    for i in range(len(segpath)-1):
      b = gfapy.SegmentEnd(segpath[i+1]).inverted()
      ls = self.segment(a.segment).end_relations(a.end_type, b, "dovetails")
      if len(ls) != 1:
        msg = "A single link was expected between {}".format(a) + \
              "and {}".format(b) + "{} were found".format(len(ls))
        raise gfapy.ValueError(msg)
      l = ls[0]
      if not l.overlap:
        cut = 0
      elif all(op.code in ["M","="] for op in l.overlap):
        cut = sum([len(op) for op in l.overlap])
      else:
        raise gfapy.ValueError(
            "Merging is only allowed if all operations are M/=")
      total_cut += cut
      last_reversed = (b.end_type == "R")
      self._add_segment_to_merged(merged, self.segment(b.segment),
          last_reversed, cut, False, enable_tracking=enable_tracking,
          merged_name=merged_name)
      a = gfapy.SegmentEnd(b).inverted()
      if self._progress:
        self._progress_log("merge_linear_paths", 0.95)
    merged.vlevel = merged_vlevel
    if isinstance(merged.name, list):
      merged.name = "_".join(merged.name)
    ortag = merged.get("or")
    if isinstance(ortag, list):
      merged.set("or", ",".join(ortag))
    if not gfapy.is_placeholder(merged.sequence):
      merged.sequence = "".join(merged.sequence)
      if self._version == "gfa1":
        if not merged.LN:
          merged.LN = len(merged.sequence)
        elif self._vlevel > 0 and merged.LN != len(merged.sequence):
          raise gfapy.InconsistencyError(
              "Computed sequence length {} ".format(len(merged.sequence))+
              "and computed LN {} differ".format(merged.LN))
    if merged.length is not None:
      for count_tag in ["KC", "RC", "FC"]:
        merged.set(count_tag, None)
    else:
      factor = 1
      if cut_counts:
        factor = merged.length / (total_cut+merged.length)
      for count_tag,count in self.__sum_of_counts(segpath,factor).items():
        merged.set(count_tag, count)
    return merged, first_reversed, last_reversed

  def __link_merged(self, merged_name, segment_end, is_reversed):
    to_disconnect = self.segment(segment_end.segment).dovetails_of_end(
                                                 segment_end.end_type)
    to_add = []
    for l in to_disconnect:
      l2 = l.clone()
      if l2.to_segment == segment_end.segment:
        l2.to_segment = merged_name
        if is_reversed:
          l2.to_orient = gfapy.invert(l2.to_orient)
      else:
        l2.from_segment = merged_name
        if is_reversed:
          l2.from_orient = gfapy.invert(l2.from_orient)
      to_add.append(l2)
    for l in to_disconnect:
      l.disconnect()
    for l in to_add:
      self.add_line(l)


